homeassistant/components/synology_dsm/backup.py

```diff
@@ -0,0 +1,217 @@
+        :return: An async iterator that yields bytes.
+        """
+        if not await self.async_get_backup(backup_id):
+            raise BackupAgentError("Backup not found")
+
+        try:
+            resp = await self._file_station.download_file(
+                path=self.path,
+                filename=f"{backup_id}.tar",
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to download backup") from err
+
+        if isinstance(resp, bool) or resp is None:
+            raise BackupAgentError("Failed to download backup")
+
+        return ChunkAsyncStreamIterator(resp)
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+
+        # upload backup.tar file first
+        try:
+            await self._file_station.upload_file(
+                path=self.path,
+                filename=f"{backup.backup_id}.tar",
+                source=await open_stream(),
+                create_parents=True,
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to upload backup") from err
+
+        # upload backup_meta.json file when backup.tar was successful uploaded
+        try:
+            await self._file_station.upload_file(
+                path=self.path,
+                filename=f"{backup.backup_id}_meta.json",
+                source=json_dumps(backup.as_dict()).encode(),
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to upload backup") from err
+
+    async def async_delete_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> None:
+        """Delete a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        """
+        if not await self.async_get_backup(backup_id):
+            return
+
+        try:
+            await self._file_station.delete_file(
+                path=self.path, filename=f"{backup_id}.tar"
+            )
+            await self._file_station.delete_file(
+                path=self.path, filename=f"{backup_id}_meta.json"
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to delete the backup") from err
+
+    async def async_list_backups(self, **kwargs: Any) -> list[AgentBackup]:
+        """List backups."""
+
+        async def _download_meta_data(filename: str) -> JsonObjectType:
+            try:
+                resp = await self._file_station.download_file(
```

Comment:
`iter_chunks` is usually a bit faster since it doesn't have to do as much copy on the data and we aren't worried about the source sending oversized chunks

Reply:
at this point, we download only the meta data files which are usually quite less than 1KByte